crates/red_knot_python_semantic/src/types.rs

```diff
@@ -271,25 +276,55 @@ fn module_type_symbols<'db>(db: &'db dyn Db) -> smallvec::SmallVec<[ast::name::N
         .collect()
 }
 
-pub(crate) fn global_symbol<'db>(
-    db: &'db dyn Db,
-    lookup: SymbolLookup,
-    file: File,
-    name: &str,
-) -> Symbol<'db> {
-    // Not defined explicitly in the global scope?
-    // All modules are instances of `types.ModuleType`;
-    // look it up there (with a few very special exceptions)
-    symbol(db, lookup, global_scope(db, file), name).or_fall_back_to(db, || {
-        if module_type_symbols(db)
-            .iter()
-            .any(|module_type_member| &**module_type_member == name)
-        {
-            KnownClass::ModuleType.to_instance(db).member(db, name)
+/// Return the symbol for a member of `types.ModuleType`.
+pub(crate) fn module_type_symbol<'db>(db: &'db dyn Db, name: &str) -> Symbol<'db> {
+    if module_type_symbols(db)
+        .iter()
+        .any(|module_type_member| &**module_type_member == name)
+    {
+        KnownClass::ModuleType.to_instance(db).member(db, name)
+    } else {
+        Symbol::Unbound
+    }
+}
+
+/// Infer the public type of a symbol (its type as seen from outside its scope).
+fn symbol<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> Symbol<'db> {
+    symbol_impl(db, scope, name, RequiresExplicitReExport::No)
+}
+
+/// Infers the public type of a symbol as seen from the same file.
+///
+/// If it's not defined explicitly in the global scope, it will look it up in `types.ModuleType`
+/// with a few very special exceptions.
+///
+/// Use [`imported_symbol`] to perform the lookup as seen from outside the file (e.g. via imports).
+pub(crate) fn global_symbol<'db>(db: &'db dyn Db, file: File, name: &str) -> Symbol<'db> {
+    symbol_impl(
+        db,
+        global_scope(db, file),
+        name,
+        RequiresExplicitReExport::No,
+    )
+    .or_fall_back_to(db, || module_type_symbol(db, name))
+}
+
+/// Infers the public type of an imported symbol.
+///
+/// Unlike [`global_symbol`], this does not fall back to looking up the symbol in
+/// `types.ModuleType`. The fallback logic needs to be handled by the caller.
```

Comment:
In which cases do I need to handle the fallback? Is it always, or is it generally fine not to handle that case when doing a cross-module lookup?

Reply:
I think we could also just have separate `builtin_symbol` and `imported_symbol` functions which each incorporate the correct module-symbols fallback for that case. It's not like the other logic in this function is extensive. And if we don't want to duplicate it, we could rename this function to something like `external_symbol_impl` (and make it private) and have both `builtin_symbol` and `imported_symbol` call it and then apply their respective fallback.

Reply:
I think then I'll update the existing `builtins_symbol` to reflect that, I was a bit hesitant on doing that but I think it should be fine. Any builtins that's imported directly from the `builtins` module should go through the `ModuleLiteralType::member` method while the lookup in the builtins scope should go through `builtins_symbol`.